### ISSUE TITLE
[promise-function-async] Functions that return a Promise value should not be forced to use the async keyword

### ISSUE THREAD

**Issue by molvqingtai** [typescript-eslint/typescript-eslint#3387]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
     "@typescript-eslint/promise-function-async": [
      "error",
      {
        "allowedPromiseNames": ["Thenable"],
        "checkArrowFunctions": true,
        "checkFunctionDeclarations": true,
        "checkFunctionExpressions": true,
        "checkMethodDeclarations": true
      }
    ]
  }
}
```

```TS
const p = new Promise<number>((resolve, reject) => {
  const n = Math.round(Math.random())
  n === 1 ? resolve(n) : reject(n)
})

// eslint-disable-next-line @typescript-eslint/promise-function-async
function basicReturn(): Promise<number> {
  return p
}

async function asyncReturn(): Promise<number> {
  // eslint-disable-next-line @typescript-eslint/return-await
  return p
}

async function asyncAwaitReturn(): Promise<number> {
  return await p
}

console.log(p === basicReturn()) // true
console.log(p === asyncReturn()) // false
console.log(p === asyncAwaitReturn()) // false

```

**Expected Result**

**Actual Result**

In the above output, only basicReturn() retains the same Promise reference. If this rule forces the async keyword, you will get unpredictable errors.

**Additional Info**
> Even though the return value of an async function behaves as if it's wrapped in a Promise.resolve, they are not equivalent.
> An async function will return a different reference, whereas Promise.resolve returns the same reference if the given value is a promise.
> It can be a problem when you want to check the equality of a promise and a return value of an async function.

Reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.1` |
| `@typescript-eslint/parser`        | `4.22.1` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.26.0` |
| `node`                             | `14.16.1` |

**Comment by bradzacher** [comment]
I'm not sure I understand the request here.

You're saying the rule `promise-function-async` shouldn't force you to mark a function that returns a promise as async?

Isn't that the entire point of the rule though?

So if you don't want the functionality, don't use the rule.

**Comment by molvqingtai** [comment]
> I'm not sure I understand the request here.
> 
> You're saying the rule `promise-function-async` shouldn't force you to mark a function that returns a promise as async?
> 
> Isn't that the entire point of the rule though?
> 
> So if you don't want the functionality, don't use the rule.

In some cases I need to use the promise as a key, but forcing the async keyword I will get a new promise reference, which will cause the code to run incorrectly.
```ts
const tasks = new Map()

const run = async (): Promise<number> => {
  const oldTask = new Promise<number>((resolve, reject) => resolve(1))
  tasks.set(oldTask, { status: 1 })
  return oldTask
}

const newTask = run()

// Due to the use of the async keyword  newTask !== oldTask
tasks.get(newTask) // undefined

// If run() does not use the async keyword
tasks.get(newTask) // { status: 1 }

```

**Comment by bradzacher** [comment]
> In some cases I need to use the promise as a key

I don't think you should ever be doing this, because it's really easy to change the promise.
`async` is just one way, but also if someone chains the promise via `.then` or `.catch`, then you'll get a new promise.

It's a really *really* fragile way of doing things.

---

> It's not a good idea to turn off this rule

I don't understand what you mean here.
This rule *only* enforces that functions that return promises be marked as async.

So if you're saying you don't want functions that return promises to be marked as async, then what reason are you using the rule for?

**Comment by molvqingtai** [comment]
> > In some cases I need to use the promise as a key
> 
> I don't think you should ever be doing this, because it's really easy to change the promise.
> `async` is just one way, but also if someone chains the promise via `.then` or `.catch`, then you'll get a new promise.
> 
> It's a really _really_ fragile way of doing things.
> 
> > It's not a good idea to turn off this rule
> 
> I don't understand what you mean here.
> This rule _only_ enforces that functions that return promises be marked as async.
> 
> So if you're saying you don't want functions that return promises to be marked as async, then what reason are you using the rule for?

Because for people with OCD, they always want to follow all the rules of `typescript-eslint` exactly.
Thanks for the suggestion, it seems that I can only turn off this rule for now.
